### ISSUE TITLE
tcp.c test fail at line 154

### ISSUE THREAD

**Issue by reqshark** [sustrik/libmill#27]
cloned master this morning at 3f55497d6c666fb47d94e2ee59c72a96a1aa0452 and gisted the result of a `make check` here: https://gist.github.com/reqshark/4ad9951010480fa5aa39

```bash
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.3.0
Thread model: posix
```

**Comment by sustrik** [comment]
Yes. More work on TCP helpers is needed. I'll look into it shortly.

**Comment by reqshark** [comment]
that assertion `(res == FDW_IN)` does not always fail. 

it appears to fail every other run. I get a pass or sometimes twice it passes before a failure. 

```c
15:11 $ make check
// ...
PASS: tests/tcp
============================================================================
Testsuite summary for libmill 0.10-beta-20-gb675ea3
============================================================================
# TOTAL: 8
# PASS:  8
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

let's check values on the line before this assert:
```c
printf("--\nprintf line 377, 
  res == %d,
  FDW_IN == %d,
  FDW_ERR == %d --",
  res, FDW_IN, FDW_ERR );
```
now output directed to test-suite.log looks like:
```c
=====================================================
   libmill 0.10-beta-20-gb675ea3: ./test-suite.log
=====================================================

# TOTAL: 8
# PASS:  7
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests/tcp
===============

Assertion failed: (res == FDW_IN), function tcprecv, file tcp.c, line 378.
--
printf line 377, res == 1, FDW_IN == 1, FDW_ERR == 4 ----
printf line 377, res == 5, FDW_IN == 1, FDW_ERR == 4 --FAIL tests/tcp (exit status: 134)

```

res is `1`, then changes to `5` before fail. `FDW_ERR` is 4.

`FDW_ERR` is not always set to a value in the first `printf line 337`
other times `FDW_ERR` value the second time is never reached, nothing from printf where we see 4 now

**Comment by sustrik** [comment]
Yes. That's connection being closed while some data are still remaining in the rx buffer. I'll try to finish the whole TCP wrapper thing this week.

**Comment by sustrik** [comment]
I've committed a little patch that should solve this particular problem. Can you try it?

Still, the whole tcp thing should be rewritten to deal with TCP errors in a sane way.

**Comment by reqshark** [comment]
Great! I'll give it a try this afternoon and report back

**Comment by reqshark** [comment]
pulled this commit 1f01165000c2e621be5da26a8c28168913689d08 and ran `make check` 15 times. 

the first 14 passed. here's what happened on 15th:
```go
./test-driver: line 107: 38975 Abort trap: 6           "$@" > $log_file 2>&1
FAIL: tests/tcp


=====================================================
   libmill 0.10-beta-20-gb675ea3: ./test-suite.log
=====================================================

# TOTAL: 8
# PASS:  7
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests/tcp
===============

Assertion failed: (rc != -1), function tcplisten, file tcp.c, line 154.
FAIL tests/tcp (exit status: 134)
```

**Comment by sustrik** [comment]
Fair enough. Looks like SO_REUSEADDR is needed. Let me fix that.

**Comment by sustrik** [comment]
Can you try now?

**Comment by reqshark** [comment]
running it

**Comment by reqshark** [comment]
issue fixed by 3612d068670b750a850a54659e611075d05669e8.
